### PR TITLE
Add DRAM tools

### PR DESCRIPTION
**Pull request by gregvonkuster**
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [ ] - This PR adds a new tool or tool collection
* [x] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

### CODE REVIEW

File: tools/dram/dram_distill.xml [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        <data name="output_product_tsv" format="tabular" label="${tool.name} on ${on_string}: product_tsv"/>
+    </outputs>
+    <tests>
+        <!--
+        These tests succeed, but the outputs are empty because the DRAM
+        database is not available in the test environment. This warning
+        is displayed in the log...
+        warnings.warn('Database does not exist at path %s' % self.description_loc)
+        ...and this exception is raised:
+        ValueError: Invalid file path or buffer object type: <class 'NoneType'
```
Comment (bgruening) [inline tools/dram/dram_distill.xml:52]: Should the tool fail in this case?
Please note that you can test for a tool error. You can define that a test should fail and then check the stderr

File: tools/dram/dram_merge_annotations.xml [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+#if $rrnas_collection:
+    && test -f 'output_dir/rrnas.tsv' && mv 'output_dir/rrnas.tsv' '$output_rrnas' || echo 'No rrnas.tsv output produced'
+#end if
+#if $trnas_collection:
+    && test -f 'output_dir/trnas.tsv' && mv 'output_dir/trnas.tsv' '$output_trnas' || echo 'No trnas.tsv output produced'
+#end if
+&& test -f 'output_dir/annotations.tsv' && mv 'output_dir/annotations.tsv' '$output_annotations' || echo 'No annotations.tsv output produced'
+    ]]></command>
+    <inputs>
+        <param name="annotations_collection" type="data_collection" format="tabular" collection_type="list" label="Collection of annotation files"/>
```
Comment (bgruening) [inline tools/dram/dram_merge_annotations.xml:54]: why is it not possible here to just take 2 single files

Is the assumption that dram_annotate is used on a collection and then produces collections? Before this tool is used.

Alternatively you could use multiple=true here

File: tools/dram/dram_neighborhoods.xml [hunk 3]
```diff
@@ -0,0 +1,133 @@
+<tool id="dram_neighborhoods" name="DRAM find neighborhoods" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="@PROFILE@">
+    <description>around genes of interest</description>
+    <macros>
+        <import>macros.xml</import>
+    </macros>
+    <expand macro="requirements"/>
+    <command detect_errors="exit_code"><![CDATA[
+#import gzip
```
Comment (bgruening) [inline tools/dram/dram_neighborhoods.xml:8]: where is this needed?

File: tools/dram/dram_neighborhoods.xml [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            <param name="distance_genes" value="10"/>
+            <output name="output_neighborhood_annotations">
+                <assert_contents>
+                    <has_size value="2046495" delta="1000"/>
+                </assert_contents>
+                <assert_contents>
+                    <has_text text="neighborhood_number"/>
+                </assert_contents>
+            </output>
+       </test>
```
Comment (bgruening) [inline tools/dram/dram_neighborhoods.xml:79]: indentation

File: tools/dram/dram_strainer.xml [hunk 5]
```diff
@@ -0,0 +1,92 @@
+<tool id="dram_strainer" name="DRAM strain annotations" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="@PROFILE@">
+    <description>down to genes of interest</description>
+    <macros>
+        <import>macros.xml</import>
+    </macros>
+    <expand macro="requirements"/>
+    <command detect_errors="exit_code"><![CDATA[
+#import gzip
```
Comment (bgruening) [inline tools/dram/dram_strainer.xml:8]: is this needed?

File: tools/dram/dram_distill.xml (hunk 1 above)
Comment (gregvonkuster) [reply tools/dram/dram_distill.xml:52]: Failing the tool would be tricky with the current DRAM version.  DRAM returns a 0 exit code when the database is missing, so we'd have to check for the database in the tool command line and fail if it's missing.

But DRAM accesses a config file stored on disk to locate the database (this is why these DRAM tools don't include data tables).  So to locate the database, we'd have to parse the output of this command which prints the config file.
```
$DRAM-setup.py print_config
```
The config file includes entries like this, so we'd have to parse the line and get the database location and then check to see if it exists.
```
KOfam db: /home/galaxy/tool-data/DRAM/1.3.5/kofam_profiles.hmm
```
To me this seems a bit brittle, but I'll do this if you feel it is best - let me know.

Also, I'll submit a ticket to the DRAM developers to return a 1 exit code if the database is missing - this is a better solution, but will  not be available until a later version.

File: tools/dram/dram_neighborhoods.xml (hunk 3 above)
Comment (gregvonkuster) [reply tools/dram/dram_neighborhoods.xml:8]: Ah, cruft from development.  Removed - thanks!

File: tools/dram/dram_neighborhoods.xml (hunk 4 above)
Comment (gregvonkuster) [reply tools/dram/dram_neighborhoods.xml:79]: Fixed - thanks!

File: tools/dram/dram_strainer.xml (hunk 5 above)
Comment (gregvonkuster) [reply tools/dram/dram_strainer.xml:8]: Removed - thanks!

File: tools/dram/dram_merge_annotations.xml (hunk 2 above)
Comment (bgruening) [reply tools/dram/dram_merge_annotations.xml:54]: ping

File: tools/dram/dram_distill.xml (hunk 1 above)
Comment (bgruening) [reply tools/dram/dram_distill.xml:52]: @gregvonkuster where is this written to?

> ValueError: Invalid file path or buffer object type: <class 'NoneType'

You can with `stdio` filter stdout and raise an error level on your own if needed. (checking with the DRAM devs is a very good idea as well.)

Where is this config-file located? Can it be in the HOME or in the JWD directory? How do you think admins should configure the tool? If *.loc file is not possible we need a README I think.

File: tools/dram/dram_merge_annotations.xml (hunk 2 above)
Comment (gregvonkuster) [reply tools/dram/dram_merge_annotations.xml:54]: > why is it not possible here to just take 2 single files
> Alternatively you could use multiple=true here

The DRAM tools are very restrictive with inputs / parameters.  This tool requires a directory for the input and a directory for the output - files do not work in either case.   I even tried implementing the wrapper without using the * wildcard on the input directory and the underlying tool failed.
```
usage: DRAM.py merge_annotations [-h] [-i INPUT_DIRS] [-o OUTPUT_DIR]

options:
  -h, --help            show this help message and exit
  -i INPUT_DIRS, --input_dirs INPUT_DIRS
                        Path with wildcards pointing to DRAM annotation output directories (default: None)
  -o OUTPUT_DIR, --output_dir OUTPUT_DIR
                        Path to output merged annotations files (default: None)
```

File: tools/dram/dram_distill.xml (hunk 1 above)
Comment (gregvonkuster) [reply tools/dram/dram_distill.xml:52]: (checking with the DRAM devs is a very good idea as well.)

I've opened this issue https://github.com/WrightonLabCSU/DRAM/issues/209

Comment (gregvonkuster) [reply tools/dram/dram_distill.xml:53]: > You can with `stdio` filter stdout and raise an error level on your own if needed. (checking with the DRAM devs is a very good idea as well.)

I've added this to dram_annotate and dram_distill, so tools and tests now fail when the database is missing.

The dram_neighborhoods and dram_strainer tools use the database only if the optional --categories param is used., but this param is broken in version 3.5.1 - https://github.com/WrightonLabCSU/DRAM/issues/194.

So I believe all tools are now covered for this.

**Pull request closed without merging**